SignalMessaging/environment/migrations/OWS109OutgoingMessageState.m

```diff
@@ -0,0 +1,81 @@
+//
+//  Copyright (c) 2018 Open Whisper Systems. All rights reserved.
+//
+
+#import "OWS109OutgoingMessageState.h"
+#import <SignalServiceKit/OWSPrimaryStorage.h>
+#import <SignalServiceKit/TSOutgoingMessage.h>
+#import <YapDatabase/YapDatabaseTransaction.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+// Increment a similar constant for every future DBMigration
+static NSString *const OWS109OutgoingMessageStateMigrationId = @"109";
+
+@implementation OWS109OutgoingMessageState
+
++ (NSString *)migrationId
+{
+    return OWS109OutgoingMessageStateMigrationId;
+}
+
+// Override parent migration
+- (void)runUpWithCompletion:(OWSDatabaseMigrationCompletion)completion
+{
+    OWSAssert(completion);
+
+    OWSDatabaseConnection *dbConnection = (OWSDatabaseConnection *)self.primaryStorage.newDatabaseConnection;
+
+    [dbConnection asyncReadWriteWithBlock:^(YapDatabaseReadWriteTransaction *_Nonnull transaction) {
+        NSMutableArray<NSString *> *messageIds =
+            [[transaction allKeysInCollection:TSOutgoingMessage.collection] mutableCopy];
```

Comment:
This is better - and probably fine for all reasonably modern devices with reasonably sized message db's.

How many keys can we hold in memory before we crash on a 5c? There's got to be some upper bound. I guess if the answer is "more than 100k", I'm not too worried about it. Let's just make sure this is in the test plan.

Reply:
+1